src/Pact/Analyze/Types.hs

```diff
@@ -646,27 +541,146 @@ data Prop a where
   -- Authorization
 
   -- | Whether a transaction contains a signature that satisfied the named key set
-  KsNameAuthorized :: KeySetName      ->                                   Prop Bool
+  KsNameAuthorized :: KeySetName      ->                                   PropSpecific Bool
   -- | Whether a row has its keyset @enforce@d in a transaction
-  RowEnforced      :: Prop TableName  -> Prop ColumnName -> Prop RowKey -> Prop Bool
+  RowEnforced      :: Prop TableName  -> Prop ColumnName -> Prop RowKey -> PropSpecific Bool
 
-pattern PAnd :: Prop Bool -> Prop Bool -> Prop Bool
-pattern PAnd a b = PLogical AndOp [a, b]
+deriving instance Eq a => Eq (PropSpecific a)
+deriving instance Show a => Show (PropSpecific a)
 
-pattern POr :: Prop Bool -> Prop Bool -> Prop Bool
-pattern POr a b = PLogical OrOp [a, b]
+class sub :<: sup where
+  inject  :: sub a -> sup a
+  project :: sup a -> Maybe (sub a)
 
-pattern PNot :: Prop Bool -> Prop Bool
-pattern PNot a = PLogical NotOp [a]
+pattern Inj :: sub :<: sup => sub a -> sup a
+pattern Inj a <- (project -> Just a) where
+  Inj a = inject a
+
+instance PropSpecific :<: Prop where
+  inject = PropSpecific
+  project = \case
+    PropSpecific a -> Just a
+    _              -> Nothing
+
+instance Numerical Prop :<: Prop where
+  inject = PureProp . Numerical
+  project (PureProp (Numerical a)) = Just a
+  project _                        = Nothing
+
+instance Numerical Invariant :<: Invariant where
+  inject = PureInvariant . Numerical
+  project (PureInvariant (Numerical a)) = Just a
+  project _                             = Nothing
+
+data PureTerm et t a where
```

Comment:
if there's a core of this change, this is it. this is the kernel shared by all three languages

Reply:
Looks great. As discussed, taking the name of `Term` into account, we should figure out what the best name is for this

Reply:
But we can merge this before coming up with the name